nixos/modules/services/web-apps/libretranslate.nix

```diff
@@ -0,0 +1,197 @@
+      };
+      extraArgs = lib.mkOption {
+        type = with lib.types; attrsOf (nullOr (oneOf [ bool str int (listOf (oneOf [ bool str int ])) ]));
+        default = { };
+        example = {
+          debug = true;
+          disable-files-translation = true;
+          url-prefix = "translate";
+        };
+        description = lib.mdDoc "Extra arguments passed to the LibreTranslate.";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    environment.systemPackages = [ ltmanageKeysCli ];
+
+    systemd.tmpfiles.rules = lib.mkIf (cfg.dataDir != "/var/lib/libretranslate") [
+      "d '${cfg.dataDir}' 0750 ${cfg.user} ${cfg.group} - -"
+      "z '${cfg.dataDir}' 0750 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    systemd.services.libretranslate = {
+      description = "LibreTranslate service";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = {
+        HOME = cfg.dataDir;
+      };
+      serviceConfig = lib.mkMerge [{
+        Type = "simple";
+        ExecStart = ''
+          ${cfg.package}/bin/libretranslate \
+            '--host' '${cfg.host}' \
+            '--port' '${toString cfg.port}' \
+            ${lib.optionalString cfg.enableApiKeys "'--api-keys'"} \
+            ${lib.optionalString cfg.disableWebUI "'--disable-web-ui'"} \
+            ${lib.optionalString cfg.updateModels "'--update-models'"} \
+            ${lib.optionalString (cfg.threads != null) "'--threads' '${toString cfg.threads}'"} \
+            ${lib.cli.toGNUCommandLineShell { } cfg.extraArgs}
+        '';
+        WorkingDirectory = cfg.dataDir;
+        User = cfg.user;
+        Group = cfg.group;
+        ProcSubset = "all";
+        ProtectProc = "invisible";
+        UMask = "0027";
+        CapabilityBoundingSet = "";
+        NoNewPrivileges = true;
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        PrivateUsers = true;
+        ProtectHostname = true;
+        ProtectClock = true;
+        ProtectKernelTunables = true;
+        ProtectKernelModules = true;
+        ProtectKernelLogs = true;
+        ProtectControlGroups = true;
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        LockPersonality = true;
+        MemoryDenyWriteExecute = false;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        RemoveIPC = true;
+        PrivateMounts = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [ "~@cpu-emulation @debug @ipc @keyring @mount @obsolete @privileged @setuid" ];
+      }
+      (lib.mkIf (cfg.dataDir == "/var/lib/libretranslate") {
+        StateDirectory = "libretranslate";
+        StateDirectoryMode = "0750";
+      })
+      (lib.mkIf (cfg.dataDir != "/var/lib/libretranslate") {
+        ReadWritePaths = cfg.dataDir;
+      })];
```

Comment:
Maybe it would be a bit easier to bindmount the dataDir to something like `/run/libretranslate` instead (or wherever else it makes sense to put a temporary working directory)?

Reply:
No, models for translation are downloaded to this folder.

Reply:
I think you misunderstood the suggestion a bit. Setting [`serviceConfig.BindPaths`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#BindPaths=) to `[ "${dataDir}:/run/libretranslate" ]` would have the effect of storing all the data in `dataDir`, while the service can find it in `/run/libretranslate`, making it so that we can get rid of the `mkIf`s and `mkMerge`. But it's not that big of a deal.